modules/core/src/lib/layer.js

```diff
@@ -551,7 +554,11 @@ export default class Layer extends Component {
   restorePickingColors() {
     const {pickingColors, instancePickingColors} = this.getAttributeManager().attributes;
     const colors = pickingColors || instancePickingColors;
-    colors.updateSubBuffer({startOffset: 0});
+    // The picking color cache may have been freed and then reallocated. This ensures we read from the currently allocated cache.
+    if (this.internalState.usesPickingColorCache) {
+      colors.value = pickingColorCache;
+    }
+    colors.updateSubBuffer({startOffset: 0, endOffset: this.getNumInstances() * 3});
```

Comment:
`this.getNumInstances()` does not necessarily match the size of the picking buffer in all layers. I suggest you keep the `pickingColorCache.subarray` logic just to keep track of the correct size of the attribute.

Reply:
For my own edification, is there an example of a layer that doesn't provide its own picking colors where `getNumInstances` would differ from the size of the picking buffer? As I'm pretty new to the deck codebase I'm unclear on how the `numInstances` provided to `calculateInstancePickingColors` would not be the same as that returned by `getNumInstances`. 

Regardless; will update to keep the subarray and read length from that as it does certainly seem safer.

Reply:
A layer does not have to be instanced; when registering an attribute, it may pass in an option `noAlloc: true` to bypass the default allocation. In that case, `pickingColor` can be any size.